### PR TITLE
Extended methods on Enumerable module

### PR DESCRIPTION
**Pull request by kirokoNjenga**
**What does this PR do**
I included My own methods in the extended Enumerable module:
#my_each
#my_each_with_index
#my_select
#my_all?
#my_any?
#my_none?
#my_count
#my_map - able to take in proc and/or a block
#my_inject
#multiply_els to test #my_inject

### CODE REVIEW

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @Samkiroko, impressive work here :+1:. However, a little change is required :fast_forward:**
- When a regex is passed to **my_any?**, it should return false only if none of the enum elements matches the regex.
`%w[boy, book, box].my_any?(/b/)` should return `true` and not `false` 

_**Kindly make this adjustment and submit another code review request.**_

**Any question? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  end
+
+  def my_any?
+    result = false
+    my_each do |i|
+      result = true if yield i
+      break if result
+    end
+    result
+  end
```
Comment (shshamim63) [inline enumerable.rb]: **`my_any?`**  identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, false, nil].any?  #should return true.
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 'class', nil].any?(Integer)  #should return true
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[hello hi howdy].any?(/d/) #should return false
%w[hello11 hi howdy].any?(/d/) #should return true
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
[1, 2, 3].any?(3)  #should return true
```
While playing with those failing cases came up with a solution as below. Had to change current my_any? with the following code and updated the any_check method(commented below). And it worked.
```ruby
  def my_any?(pattern = nil)
    my_each do |x|
      if block_given?
        return true if yield x
      elsif !pattern.nil?
        return true if any_check(pattern, x)
      elsif x
        return true
      end
    end
    false
  end
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+def any_check(pattern, exponent)
+  if pattern.is_a? Regexp
+    return false if pattern =~ exponent
+  elsif pattern.is_a? Class
+    return true if exponent.is_a? pattern
+  elsif exponent == pattern
+    return true
+  end
+  false
+end
```
Comment (shshamim63) [inline enumerable.rb]:
```ruby
def any_check(pattern, exponent)
  if pattern.is_a? Regexp
    return true if pattern =~ exponent
  elsif pattern.is_a? Class
    return true if exponent.is_a? pattern
  elsif exponent == pattern
    return true
  end
  false
end
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :space_invader: Changes Required :space_invader: 
___
Hi @Samkiroko :wave:,
Awesome job. Almost done with this project.  :clap: . But identified some issues that are blocking me to approve this project now. Please see the comments below to complete this project.
___
### :spades: Mandatory Changes :spades: 
___
1. Your my_any? is not covering some edge cases. I have marked them below with a sample solution. Please fix the issue.
___
### :clubs: Optional Changes :clubs: 
___
1. I can not see a `multiply_els` that uses `my_inject` as described on Odin ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/multiply_els.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). Remember this method is not an enumerable method. So should be placed outside of the module. Extremely sorry that the previous reviewer missed it to mention. So I am marking it as optional.
___
##### Please fix the above-mentioned issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Twitter](https://twitter.com/Shshamim090)

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @Samkiroko      
#### Great Job so far, but there are a few more things required. 


- [ ] `my_any?`:

- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p [12,3,34].my_any?(Integer)` should return `true`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild'].my_any?(/z/)` should return `false`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['cat',2,32,].my_any?('cat')` should return `true`




### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by kirokoNjenga** [comment]
> ## Hi @Samkiroko 
> #### Great Job so far, but there are a few more things required.
> * [ ]  `my_any?`:
> * when no block or argument is given returns true if at least one of the collection is not false or nil
>   `p [nil, false, nil, false].my_any?` - should return `false`
> * when a class is passed as an argument returns true if at least one of the collection is a member of such class
>   `p [12,3,34].my_any?(Integer)` should return `true`
> * when a Regex is passed as an argument returns false if none of the collection matches the Regex
>   `p ['ghetto','wild'].my_any?(/z/)` should return `false`
> * when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
>   `p ['cat',2,32,].my_any?('cat')` should return `true`
> 
> ### Make changes and submit another code review request.
> _Happy coding_
> (Any questions Feel free to contact me on slack @kobiyoyo)

There is no standard way of checking this method every requires a different return

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @Samkiroko      
#### Great Job so far, you are almost there ,just few more things required. 
- [ ] Kindly refer to this doc for guidance https://ruby-doc.org/core-2.6.5/Enumerable.html 

- [ ] `my_any?`:

- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p [12,3,34].my_any?(Integer)` should return `true`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild'].my_any?(/z/)` should return `false`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['cat',2,32,].my_any?('cat')` should return `true`
- [ ] Am getting this error `undefined method any_check`



### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by kirokoNjenga** [comment]
**What does this PR do**
- Message to the TSC who will go through the code, take time to see the previous comments
- I have been back and forth because every tsc require different result on the same code its seem there is no standard way to test. I have done my test case please base your evaluation on that.
- In case you fill something is not correct I will appreciate the code on how you would do it.
- Think I also need to move on, Am wasting my partner time.
- Am sorry but am frustrated  with the reviews I have received so far

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Require changes🛑
# 🙋‍♂️ Samuel,
> Sorry for the frustration. Don't worry. This project can be a bit challenging. The examples used in the previous reviews converge in the same direction. It easier if you take ruby documentation on this method, get their functionalities and try to replicate them. Aside from ```my_map``` none of the methods has extra functionalities.
You have come a long way so far.  Almost all the methods work fine except ```my_map```, it is not accepting proc as you mentioned in your pull request.
* [ ] should have working ```my_map```  identical and returns the same thing as ruby's 
[`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)).
refer to the Odin project link above, for this particular requirement.
For example:
``` my_proc = Proc.new {|num| num < 3 } ```
``` [1,2,3,4].my_map(my_proc){|num| num < 3 }```  should return ```# [true, true, false, false]```
refer to the Odin project link above
Kindly make this change and submit the project for review🙂.
#### Happy coding💻
Addo

**Comment by kirokoNjenga** [comment]
- Thank to know that  the other methods are correct 
- Done change on the my_map as recommended
- Hope I can move on

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello @Samkiroko  ,  You have done a great job with this project. :clap:

**Pull request merged**